dbal/dsn.go

```diff
@@ -15,7 +15,11 @@ const (
 // - shared but unique in the same process
 // see: https://sqlite.org/inmemorydb.html
 func IsMemorySQLite(dsn string) bool {
-	r := regexp.MustCompile(`sqlite://file::?\w+:\?_fk=true&*(cache=shared)?(mode=memory)?`)
+	if dsn == "memory" {
+		return true
+	}
+
+	r := regexp.MustCompile(`sqlite://(file)?::?\w+:\?_fk=true&*(cache=shared)?(mode=memory)?`)
```

Comment:
This pattern still does not fully convince me.
If `file` is optional, shoudn't the colon also be optional?
```suggestion
	r := regexp.MustCompile(`sqlite://(file:)?:?\w+:\?_fk=true&*(cache=shared)?(mode=memory)?`)
```
The query parameters are only matching in the order they are specified here, so
```
sqlite://file:uniquedb:?_fk=true&mode=memory&cache=shared
^-- match start                            ^-- match end
sqlite://file:uniquedb:?_fk=true&cache=shared&mode=memory
^-- match start                                         ^-- match end
```
Also, we are missing a `&` in between the query params, and why is it `*` after the first `&`?
```suggestion
	r := regexp.MustCompile(`sqlite://(file)?::?\w+:\?_fk=true(&cache=shared)?(&mode=memory)?`)
```
Also, don't we have to have either the file `memory` or some other file name AND the `mode=memory`?

Reply:
refactoring this

Reply:
so the golang regex engine does not have positive lookahead (which seems to be required for matches in any position).
Do you know of any other methodology? https://stackoverflow.com/questions/4389644/regex-to-match-string-containing-two-names-in-any-order/4389683#4389683

Reply:
Either `...(&cache=shared)?(&mode=memory)?(&cache=shared)?` or `url.Parse` and ditch the regex completely.

Reply:
Or split the string into query and rest, do a regex on rest and parse the query string to do checks on that.